src/coreclr/jit/gentree.cpp

```diff
@@ -19791,6 +19791,39 @@ uint16_t GenTreeLclVarCommon::GetLclOffs() const
     }
 }
 
+#if defined(TARGET_XARCH) && defined(FEATURE_HW_INTRINSICS)
+//------------------------------------------------------------------------
+// GetOverwrittenOpNumForFMA: check if the result is written into one of the operands
+//
+// Return Value:
+//     The operand number or 0 if not overwritten.
+//
+unsigned GenTreeHWIntrinsic::GetOverwrittenOpNumForFMA(GenTree* use, GenTree* op1, GenTree* op2, GenTree* op3)
```

Comment:
After thinking about this, I wonder if its correctly taking everything into account.

That is, today (without this PR) we only take into account which operand is contained for picking the instruction form. This in general removes a `movups` by folding the memory read into the instruction.

The issue tracked by https://github.com/dotnet/runtime/issues/12984 is that there are cases where the destination register isn't ideal and so it results in an additional `vmovaps`.

This check partially accounts for that by checking if any of the operands is also the storage location but there are many scenarios that this won't be the case and where it may be a different local entirely, that is it helps cover:
* The register allocator should preference the overwritten operand.

It does not however cover:
* Lowering should bias the containment such that the overwritten operand is a last use if possible.

That is, imagine the cases discussed here: https://github.com/dotnet/runtime/pull/58196#discussion_r702203634. In this thread, its pointed out the even though the C# code was `c = Fma.MultiplyAdd(c, a, Sse.LoadVector128((float*)&b));` and so `op3` should be contained and `op1` (`c`) should be last use, they are actually different locals (it gets a temp, `t19`, rather than the same variable slot, `V02`). This should still be accountable/detectable by looking at and determining that `op1` is `lastUse`

Reply:
I think we can cover this by still biasing the most towards the overwritten op, but also by making an additional bias towards a last use operand when none of the ops are being overwritten.

Reply:
So, we preference a form based on both ```which operands are contained``` and ```which operands are last use``` in the case ```overwrittenOpNum == 0``` ?

Reply:
Yes, I think that would cover things and address the scenarios like you had called out earlier in the PR.